ArduCopter/mode.cpp

```diff
@@ -287,14 +287,16 @@ bool Copter::set_mode(Mode::Number mode, ModeReason reason)
         return false;
     }
 
+    // perform any cleanup required by previous flight mode
+    exit_mode(flightmode, new_flightmode);
+
     if (!new_flightmode->init(ignore_checks)) {
         mode_change_failed(new_flightmode, "initialisation failed");
+        // return to previous flight mode doing proper initialization
+        flightmode->init(true);
         return false;
     }
```

Comment:
I fear unforeseen consequences with this one. At the least is is quite a big change in behavior to re-enter the current mode. RTL will do climb min again for example. I suspect its the ones that we don't think of that will catch us tho....

Reply:
Removed